// graphql-toolkit/dist/commonjs/loaders/load-typedefs.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const graphql_1 = require("graphql");
const isValidPath = require("is-valid-path");
const isGlob = require("is-glob");
const glob_1 = require("glob");
const valid_url_1 = require("valid-url");
const load_from_url_1 = require("./load-from-url");
const path_1 = require("path");
const load_from_json_file_1 = require("./load-from-json-file");
const load_from_gql_file_1 = require("./load-from-gql-file");
const load_from_code_file_1 = require("./load-from-code-file");
const debugLog_1 = require("../utils/debugLog");
const fix_windows_path_1 = require("../utils/fix-windows-path");
const GQL_EXTENSIONS = ['.gql', '.graphql', '.graphqls'];
const CODE_FILE_EXTENSIONS = ['.ts', '.tsx', '.js', '.jsx'];
function filterFiles(files) {
    return files.filter(file => !file.includes('node_modules') && !file.endsWith('.d.ts') && !file.endsWith('.spec.ts') && !file.endsWith('.spec.js') && !file.endsWith('.test.ts') && !file.endsWith('.test.js'));
}
async function loadTypedefs(pointToSchema, options = {}, filterKinds = [], cwd = process.cwd()) {
    const schemasPaths = normalizeSchemaString(pointToSchema);
    let found = [];
    for (const schemaPath of schemasPaths) {
        if (isSchemaString(schemaPath)) {
            found.push({
                filePath: schemaPath,
                content: graphql_1.parse(schemaPath),
            });
        }
        else if (!valid_url_1.isUri(schemaPath)) {
            const fixedPath = fix_windows_path_1.fixWindowsPath(schemaPath);
            if (isValidPath(fixedPath) || isGlob(fixedPath)) {
                const relevantFiles = filterFiles(!isGlob(fixedPath)
                    ? [fixedPath]
                    : glob_1.sync(fixedPath, {
                        cwd,
                        ignore: options.ignore || [],
                        absolute: true,
                    }));
                found.push(...(await Promise.all(relevantFiles.map(async (p) => ({ filePath: p, content: await loadSingleFile(p, { noRequire: options.noRequire, tagPluck: options.tagPluck || {} }, cwd) })))));
            }
        }
        else if (valid_url_1.isUri(schemaPath)) {
            found.push({
                filePath: schemaPath,
                content: await load_from_url_1.loadFromUrl(schemaPath, options),
            });
        }
    }
    let allFoundDocuments = graphql_1.concatAST(found.map(a => a.content).filter(a => a));
    if (allFoundDocuments.definitions.length > 0 && filterKinds && filterKinds.length > 0) {
        const invalidDefinitions = allFoundDocuments.definitions.filter(d => filterKinds.includes(d.kind));
        if (invalidDefinitions.length > 0) {
            invalidDefinitions.forEach(d => {
                debugLog_1.debugLog(`Filtered document of kind ${d.kind} due to filter policy (${filterKinds.join(', ')})`);
            });
        }
        found = found.map(documentFile => ({
            filePath: documentFile.filePath,
            content: {
                kind: graphql_1.Kind.DOCUMENT,
                definitions: documentFile.content ? documentFile.content.definitions.filter(d => !filterKinds.includes(d.kind)) : null,
            },
        }));
    }
    const nonEmpty = found.filter(f => f.content && f.content.definitions && f.content.definitions.length > 0);
    if (nonEmpty.length === 0) {
        throw new Error(`Unable to find any GraphQL type defintions for the following pointers: ${schemasPaths.join(', ')}`);
    }
    return nonEmpty;
}
exports.loadTypedefs = loadTypedefs;
async function loadSingleFile(filePath, options = {}, cwd = process.cwd()) {
    const extension = path_1.extname(filePath).toLowerCase();
    const fullPath = fix_windows_path_1.fixWindowsPath(path_1.isAbsolute(filePath) ? filePath : path_1.resolve(cwd, filePath));
    try {
        if (extension === '.json') {
            return await load_from_json_file_1.loadFromJsonFile(fullPath);
        }
        else if (GQL_EXTENSIONS.includes(extension)) {
            return await load_from_gql_file_1.loadFromGqlFile(fullPath);
        }
        else if (CODE_FILE_EXTENSIONS.includes(extension)) {
            return await load_from_code_file_1.loadFromCodeFile(fullPath, options);
        }
    }
    catch (e) {
        debugLog_1.debugLog(`Failed to find any GraphQL type definitions in: ${filePath} - ${e.message}`);
        return null;
    }
    return null;
}
exports.loadSingleFile = loadSingleFile;
function isSchemaString(str) {
    // XXX: is-valid-path or is-glob treat SDL as a valid path
    // (`scalar Date` for example)
    // this why checking the extension is fast enough
    // and prevent from parsing the string in order to find out
    // if the string is a SDL
    if (/\.[a-z0-9]+$/i.test(str)) {
        return false;
    }
    try {
        graphql_1.parse(str);
        return true;
    }
    catch (e) {
        return false;
    }
}
function normalizeSchemaString(str) {
    if (Array.isArray(str)) {
        return str;
    }
    return [str];
}
