// @nestjs/graphql/dist/graphql.factory.js
"use strict";
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
var __metadata = (this && this.__metadata) || function (k, v) {
    if (typeof Reflect === "object" && typeof Reflect.metadata === "function") return Reflect.metadata(k, v);
};
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
Object.defineProperty(exports, "__esModule", { value: true });
const common_1 = require("@nestjs/common");
const apollo_server_core_1 = require("apollo-server-core");
const graphql_tools_1 = require("graphql-tools");
const fs_1 = require("fs");
const graphql_1 = require("graphql");
const lodash_1 = require("lodash");
const graphql_ast_explorer_1 = require("./graphql-ast.explorer");
const graphql_schema_builder_1 = require("./graphql-schema-builder");
const delegates_explorer_service_1 = require("./services/delegates-explorer.service");
const resolvers_explorer_service_1 = require("./services/resolvers-explorer.service");
const scalars_explorer_service_1 = require("./services/scalars-explorer.service");
const extend_util_1 = require("./utils/extend.util");
const remove_temp_util_1 = require("./utils/remove-temp.util");
let GraphQLFactory = class GraphQLFactory {
    constructor(resolversExplorerService, delegatesExplorerService, scalarsExplorerService, graphqlAstExplorer, gqlSchemaBuilder) {
        this.resolversExplorerService = resolversExplorerService;
        this.delegatesExplorerService = delegatesExplorerService;
        this.scalarsExplorerService = scalarsExplorerService;
        this.graphqlAstExplorer = graphqlAstExplorer;
        this.gqlSchemaBuilder = gqlSchemaBuilder;
    }
    mergeOptions(options = { typeDefs: [] }) {
        return __awaiter(this, void 0, void 0, function* () {
            const resolvers = this.resolversExplorerService.explore();
            const typesResolvers = extend_util_1.extend(this.scalarsExplorerService.explore(), resolvers);
            const transformSchema = (schema) => __awaiter(this, void 0, void 0, function* () { return options.transformSchema ? yield options.transformSchema(schema) : schema; });
            if (options.autoSchemaFile) {
                const autoGeneratedSchema = yield this.gqlSchemaBuilder.build(options.autoSchemaFile, options.buildSchemaOptions, this.resolversExplorerService.getAllCtors());
                const executableSchema = graphql_tools_1.makeExecutableSchema({
                    resolvers: extend_util_1.extend(typesResolvers, options.resolvers),
                    typeDefs: apollo_server_core_1.gql `
          ${graphql_1.printSchema(autoGeneratedSchema)}
        `,
                    resolverValidationOptions: Object.assign(Object.assign({}, (options.resolverValidationOptions || {})), { requireResolversForResolveType: false }),
                });
                let schema = options.schema
                    ? graphql_tools_1.mergeSchemas({
                        schemas: [options.schema, executableSchema],
                    })
                    : executableSchema;
                const autoGeneratedSchemaConfig = autoGeneratedSchema.toConfig();
                const schemaConfig = this.overrideOrExtendResolvers(schema.toConfig(), autoGeneratedSchemaConfig);
                schema = new graphql_1.GraphQLSchema(schemaConfig);
                return Object.assign(Object.assign({}, options), { typeDefs: undefined, schema: yield transformSchema(schema) });
            }
            if (lodash_1.isEmpty(options.typeDefs)) {
                return Object.assign(Object.assign({}, options), { typeDefs: undefined, schema: yield transformSchema(options.schema) });
            }
            const executableSchema = graphql_tools_1.makeExecutableSchema({
                resolvers: extend_util_1.extend(typesResolvers, options.resolvers),
                directiveResolvers: options.directiveResolvers,
                schemaDirectives: options.schemaDirectives,
                typeDefs: apollo_server_core_1.gql `
        ${options.typeDefs}
      `,
                resolverValidationOptions: options.resolverValidationOptions,
            });
            const schema = options.schema
                ? graphql_tools_1.mergeSchemas({
                    schemas: [options.schema, executableSchema],
                })
                : executableSchema;
            remove_temp_util_1.removeTempField(schema);
            return Object.assign(Object.assign({}, options), { typeDefs: undefined, schema: yield transformSchema(schema) });
        });
    }
    overrideOrExtendResolvers(executableSchemaConfig, autoGeneratedSchemaConfig) {
        const schemaConfig = autoGeneratedSchemaConfig;
        const rootResolverKeys = [
            'mutation',
            'query',
            'subscription',
        ];
        rootResolverKeys
            .filter(key => executableSchemaConfig[key] && schemaConfig[key])
            .forEach(key => {
            const executableSchemaFields = executableSchemaConfig[key].getFields();
            const schemaFields = schemaConfig[key].getFields();
            lodash_1.forEach(executableSchemaFields, (value, resolverName) => {
                if (schemaFields[resolverName]) {
                    schemaFields[resolverName].resolve =
                        executableSchemaFields[resolverName].resolve;
                    if (schemaFields[resolverName].subscribe) {
                        schemaFields[resolverName].subscribe =
                            executableSchemaFields[resolverName].subscribe;
                    }
                }
                else {
                    schemaFields[resolverName] = executableSchemaFields[resolverName];
                }
            });
        });
        const getAutoGeneratedTypeByName = (name) => schemaConfig.types.find(type => type.name === name);
        executableSchemaConfig.types
            .filter(type => type instanceof graphql_1.GraphQLObjectType)
            .forEach((type) => {
            const fields = type.getFields();
            lodash_1.forEach(fields, (value, key) => {
                if (!value.resolve) {
                    return;
                }
                const autoGeneratedType = getAutoGeneratedTypeByName(type.name);
                if (!autoGeneratedType) {
                    return;
                }
                const field = autoGeneratedType.getFields()[key];
                field && (field.resolve = value.resolve);
            });
        });
        return schemaConfig;
    }
    createDelegates() {
        return this.delegatesExplorerService.explore();
    }
    generateDefinitions(typeDefs, options) {
        return __awaiter(this, void 0, void 0, function* () {
            if (lodash_1.isEmpty(typeDefs) || !options.definitions) {
                return;
            }
            const tsFile = yield this.graphqlAstExplorer.explore(apollo_server_core_1.gql `
        ${typeDefs}
      `, options.definitions.path, options.definitions.outputAs);
            if (!fs_1.existsSync(options.definitions.path) ||
                !fs_1.lstatSync(options.definitions.path).isFile() ||
                fs_1.readFileSync(options.definitions.path, 'utf8') !== tsFile.getFullText()) {
                yield tsFile.save();
            }
        });
    }
};
GraphQLFactory = __decorate([
    common_1.Injectable(),
    __metadata("design:paramtypes", [resolvers_explorer_service_1.ResolversExplorerService,
        delegates_explorer_service_1.DelegatesExplorerService,
        scalars_explorer_service_1.ScalarsExplorerService,
        graphql_ast_explorer_1.GraphQLAstExplorer,
        graphql_schema_builder_1.GraphQLSchemaBuilder])
], GraphQLFactory);
exports.GraphQLFactory = GraphQLFactory;
